### ISSUE TITLE
Custom table for subscriptions

### ISSUE THREAD

**Issue by mindctrl** [restrictcontentpro/restrict-content-pro#1711]
I'm breaking this out separate from #774.

This may be updated, but the current schema proposal is:

Field | Type | Null | Key | Default | Extra
-- | -- | -- | -- | -- | --
id | int unsigned | NO | PRI | NULL | auto_increment
object_id | bigint unsigned | NO | MUL | NULL |
object_type | varchar(20) | NO |   | 'subscription' | 
user_id | int unsigned | NO | MUL | NULL |
currency | varchar(20) | NO | | USD |
initial_amount | mediumint | NO | | NULL |
recurring_amount | mediumint unsigned | NO | | NULL |
created_date | datetime | NO | | CURRENT_TIMESTAMP |
trial_end_date | datetime | YES | | NULL |
cancellation_date | datetime | YES | | NULL |
expiration_date | datetime | YES | | NULL |
times_billed unsigned | smallint | NO | NO | 1 | |
status | varchar(12) | NO | MUL | active |
payment_profile_id | tinytext | YES | | NULL |
gateway | tinytext | NO | | NULL |

**Comment by nosegraze** [comment]
Is it worth renaming `payment_profile_id` to something a little more descriptive? Like `gateway_subscription_id`? Or `gateway_profile_id`? Just something to more clearly designate that it's a gateway value.

Also, how does `expiration_date` (and probably `cancellation_date` too) work? For example, say a subscription expires, so we fill out `expiration_date`, but then it gets renewed and reactivated. Is `expiration_date` still there or do we wipe it out? Then do we overwrite the value if the subscription expires yet again?

Also what about a column for `auto_renew` or `recurring`?

Also, now that I'm thinking about auto renew, I think we once vaguely discussed removing the `cancelled` status completely and instead thinking of cancellations more in terms of _turning off auto renew_ rather than cancelling membership. Thoughts on that?

**Comment by nosegraze** [comment]
Lol don't mind me. Yesterday I seem to have forgotten how expiration dates work.

**Comment by nosegraze** [comment]
I think I thought `times_billed` was going to be the current number of times the subscription _has been_ billed - not the number of times a subscription _will be_ billed. So you sign up, that's `1`, then you renew and that makes it `2`, etc.

**Comment by mindctrl** [comment]
@nosegraze I've been thinking the same as you re: `times_billed`. If it's not going to reflect the name of the column, we need to change it. Having a recurring subscription billed 10 times with a `times_billed` column value of `0` won't make sense.

**Comment by nosegraze** [comment]
I would definitely like to keep a count of how many times the user has been
billed so far. I think that's important for payment plans.

On 8 Mar 2018 15:00, "John Parris" <notifications@github.com> wrote:

@nosegraze <https://github.com/nosegraze> I've been thinking the same as
you re: times_billed. If it's not going to reflect the name of the column,
we need to change it. Having a recurring subscription billed 10 times with
a times_billed column value of 0 won't make sense.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/restrictcontentpro/restrict-content-pro/issues/1711#issuecomment-371512361>,
or mute the thread

**Comment by pippinsplugins** [comment]
Oops, right, `times_billed` isn't right for what I said.

I think we should have a second, related column then. In EDD we call it `bill_times`.

**Comment by nosegraze** [comment]
Also, as per Slack, we talked about this being a _memberships_ table rather than a _subscriptions_ table.

**Comment by nosegraze** [comment]
Do we also want to add a meta table at the same time?

**Comment by mindctrl** [comment]
Unless we have a specific need for it, I vote no. If we do have a need for it, we may want to consider if that data should be a column in the subscriptions table.

We added payment meta and still don't use it. The one plugin I know of that does use it uses it wrong.

**Comment by nosegraze** [comment]
I was mostly thinking if add-ons might use it, particularly when we add multiple subscriptions. For example, in Site Creation I could imagine the site ID for a subscription being stored in the subscription meta table, or in Group Accounts we might store the group ID with the subscription.

But I'm fine with omitting it for now until we actually start planning exactly how add-ons will work with that.

**Comment by tnorthcutt** [comment]
@nosegraze is there any chance more than one discount could be used for a subscription (whether now or in the future)? For instance, is it possible that a member could use a one-time discount when signing up, and then later on use another one-time discount to get a month free or partially free?

If so, perhaps a `discounts_used` table (or other naming) would make sense, which would store a `discount_id` and a `subscription_id`.

If each subscription will only ever have the possibility of using one discount, then this wouldn't matter.

**Comment by nosegraze** [comment]
Yes I guess that would be possible with manual renewals.

**Comment by mindctrl** [comment]
We store the discount code used on the payment records, which is technically where they apply right now. Storing it in two places feels wrong.

Related to this, we have another data issue to consider. Let's chat about it. Here's my current thinking (which may be cloudy, not feeling great today).

We've previously discussed memberships vs subscriptions for the table name here. If this is legitimately a record of only recurring subscriptions, it should be called subscriptions. If we're using it to also store data about non-recurring memberships, we should call it memberships.

Let's assume for now we're calling it memberships, which is how I'll refer to it below.

When we added `object_id` and `object_type` to the payments table, we chose to use the default word of `subscription` (to refer to levels) as the type. I now think that was a poor word choice, because not all payments are for subscriptions (of the recurring type), some are for one-time payments, and the lack of solid terminology is confusing.

Aside from the confusing terminology, I think it's the wrong data relationship as we move forward. Here's why.

We need a solid way to relating payment records with membership records (this table). The payment <-> plan relationship in the object_* columns doesn't work for that.

With the new membership table, when we create payment records the `object_id` should link back to the membership ID in the memberships table, not the level like it currently does. After all, we'll need to easily see all payments for a membership. There's no better way than the ID. We should link to the level in the memberships table with the `object_type` value as 'level' or 'plan' or something.

Then we'd have a clear link between payment <-> membership/subscription <-> plan (or other items in the future. a plan is just a line item.)

During the 3.0 upgrade, we can adjust the `object_id` in payments to point to the membership ID in the new membership table.

And for the #1174 mentioned above, it would not be difficult to get the memberships that used a discount code in this scenario. You can query the payments table for payments with that discount code, and get the subscription IDs there.

I still need to explore this thought more when I'm feeling better, but curious to hear your thoughts.

**Comment by nosegraze** [comment]
Yes I've been thinking this exact same thing.

We should have a `memberships` table that covers all kind of memberships, regardless of whether there's a subscription.

The payment records need to link to the memberships table.

**Comment by nosegraze** [comment]
Another column to think about: Stripe and Braintree both use `rcp_merchant_subscription_id` for storing the subscription ID in addition to the payment profile ID field for storing the gateway customer ID. Perhaps we need another column for that too. Maybe we should have:

`gateway_customer_id` (currently payment profile ID)
`gateway_subscription_id` (currently merchant subscription ID)

**Comment by nosegraze** [comment]
Thinking about #766.

We will be storing `times_billed` in the memberships table, which is essentially the number of payments that have been made so far. Do we also want to store something to designate the toatl number of times to bill (for payment plans)? So we might have 2 times billed out of 4 total payments (where `2` is `times_billed` and `4` is saved under something else that I can't think of a great name for right now :P ). This information will already be saved in the subscription level, but do we want to copy it over to the subscription in case the subscription level gets edited? Or just always get that value from the subscription level setting?

**Comment by mindctrl** [comment]
Yeah, I think we'll need to store it with the membership row here, in cases where someone changes the level settings after someone has already purchased.
